include/pybind11/pybind11.h

```diff
@@ -1173,9 +1173,16 @@ class module_ : public object {
             py::module_ m3 = m2.def_submodule("subsub", "A submodule of 'example.sub'");
     \endrst */
     module_ def_submodule(const char *name, const char *doc = nullptr) {
-        std::string full_name
-            = std::string(PyModule_GetName(m_ptr)) + std::string(".") + std::string(name);
-        auto result = reinterpret_borrow<module_>(PyImport_AddModule(full_name.c_str()));
+        const char *this_name = PyModule_GetName(m_ptr);
+        if (this_name == nullptr) {
+            throw error_already_set();
+        }
+        std::string full_name = std::string(this_name) + "." + name;
+        handle submodule = PyImport_AddModule(full_name.c_str());
+        if (!submodule) {
+            throw error_already_set();
+        }
+        auto result = reinterpret_borrow<module_>(submodule);
```

Comment:
For reference, wouldn't you also been able to do? 
```
if(!result){
            throw error_already_set();
}
```
This looks good either way though.

Reply:
IIU your question correctly, that would assume `reinterpret_borrow<module_>` is OK with `nullptr`, and does not care about the active error. While that is usually true, what if someone temporarily adds debugging code to `reinterpret_borrow`, or the `module_` constructor in this case? Their fault, of course, but why build a trap like that?
Generally, I think it is bad style to separate error handling from the source of the error more than necessary.